### DISCUSSION TITLE
Lesson%201%3A%20Hello%20Metal/2.%20Hello%20Triangle/

### DISCUSSION THREAD

**Discussion by giscus** [wmarti/MetalTutorial#4] [Announcements]
# Lesson%201%3A%20Hello%20Metal/2.%20Hello%20Triangle/

https://metaltutorial.com/Lesson%201%3A%20Hello%20Metal/2.%20Hello%20Triangle/

**Comment by romanWSgit** [comment]
Great Tutorial so far, 

but...

Maybe I forgot to import some Library or something, but "simd::float3" is marked as an undeclared identifier in 
void MTLEngine::createTriangle()
{
  simd::float3 triangleVertices[] =
  {
    {-0.5f, -0.5f, 0.0f},
    { 0.5f, -0.5f, 0.0f},
    { 0.0f,  0.5f, 0.0f}
  };

  triangleVertexBuffer = metalDevice->newBuffer(&triangleVertices,
                                                sizeof(triangleVertices),
                                                MTL::ResourceStorageModeShared);
}

wich is in located in mtl_engine.mm in my case.

Kind of stuck here... 
Help would be appreciated! 

Regards, Roman

**Reply by romanWSgit** [reply]
I included "#include <simd/simd.h>" in mil_engine.mm and this seems to work. Maybe I overlooked this.
Nevertheless ...great tutorial!

**Reply by wmarti** [reply]
Nope, seems like I actually overlooked that! Nice catch :). I'll update the docs right now. Just note that the site might go offline for a minute or so when I do, GitHub pages likes to mess everything up...

**Reply by wmarti** [reply]
Docs have been updated! Do let me know if you find anything else I've missed, I'll be sure to fix it right away :)

**Reply by romanWSgit** [reply]
Thanks very much for the fast reply.

**Comment by MERL10N** [comment]
With regards to the objective-c code needed in MTLEngine, you can abstract them away so that MTLEngine and main can be .cpp instead of .mm


First, replace @autoreleasepool in the run() function with NS::AutoReleasePool* pPool
(You can declare it in the run function or in MTLEngine.hpp as I did)

>void MTLEngine::run()
 {
            while (!glfwWindowShouldClose(glfwWindow)) {
            pPool = NS::AutoreleasePool::alloc()->init();
            metalDrawable = layer->nextDrawable();
            draw();
            pPool->release();
            glfwPollEvents();
        }    
}

Note that you can make this change in main.cpp as well

Next, declare in MTLEngine.hpp

>CA::MetalLayer* layer;


and add these lines in the init() function. Don't compile yet because there will still be errors.
>layer = CA::MetalLayer::layer();
layer->setDevice(metalDevice);
layer->setPixelFormat(MTL::PixelFormatBGRA8Unorm);
layer->setDrawableSize(CGSizeMake(width, height));
GLFWBridge::AddLayerToWindow(glfwWindow, layer);
>

Thirdly,
remove these in MTLEngine.hpp
> #include <Metal/Metal.h>  
> #include <GLFW/glfw3native.h>
> #define GLFW_EXPOSE_NATIVE_COCOA

> #include <QuartzCore/CAMetalLayer.h>


then go ahead and include

 >#include <QuartzCore/CAMetalLayer.hpp> 

Note that you should still keep #include <GLFW/glfw3.h> in MTLEngine

Lastly, this final step is based on what is found on stack overflow: https://stackoverflow.com/questions/76776751/how-to-use-metal-cpp-with-glfw

I named my file GLFWBridge.h/.mm


>#ifndef GLFWBridge_h
#define GLFWBridge_h
struct GLFWwindow;

>namespace CA
{
    class MetalLayer;
}
namespace GLFWBridge
{
    void AddLayerToWindow(GLFWwindow* window, CA::MetalLayer* layer);
}
#endif /* GLFWBridge_h */

>#include "GLFWBridge.h"
#define GLFW_EXPOSE_NATIVE_COCOA
#include <GLFW/glfw3.h>
#include <GLFW/glfw3native.h>

>#include <Metal/Metal.h>
#include <QuartzCore/QuartzCore.h>

>namespace GLFWBridge
{
    void AddLayerToWindow(GLFWwindow* window, CA::MetalLayer* layer)
    {
        NSWindow* cocoa_window = glfwGetCocoaWindow(window);
        CAMetalLayer* native_layer = (__bridge CAMetalLayer*)layer;
        [[cocoa_window contentView] setLayer:native_layer];
        [native_layer setMaximumDrawableCount:2];
        [[cocoa_window contentView] setWantsLayer:YES];
        [[cocoa_window contentView] setNeedsLayout:YES];
    }
}

Back in MTLEngine.mm


Go ahead and add this
>#include "GLFWBridge.h"

and this line in init() after layer->setDrawableSize(CGSizeMake(width, height));

> GLFWBridge::AddLayerToWindow(glfwWindow, layer);


After all these, you can go ahead and switch MTLEngine.mm to MTLEngine.cpp, as well as main.mm to main.cpp.

**Reply by MissBidule** [reply]
FYI, I did your solution and it worked great... except at the 3D step, it made my cube turn all laggy so I had to delete this line : [native_layer setMaximumDrawableCount:2] as its default value is 3.
I tried to find why it should be 3 and not 2 but did not understand...

**Reply by kptr-juku** [reply]
Thanks, I was experimenting with replacing the MetalLayer with the CPP one myself, but I could not exactly 'bridge' it :)
 I would add, that there is a typo in the first placement of `NS::AutoReleasePool* pPool` for the mtl_engine.h, it should be without capital 'R' - took me some time to figure this out :)
Also pixel format getter from layer should now be:
`MTL::PixelFormat pixelFormat = (MTL::PixelFormat)layer->pixelFormat();`

**Comment by AlexanderDevaikinEnscape** [comment]
Shouldn't we have released vertex and fragment shader objects because they were created using function with the `new` prefix in name?

```
vertexShader->release();
fragmentShader->release();
```

**Comment by shi-yan** [comment]
you forgot to release vertexShader and fragmentShader?

**Comment by chuzcjoe** [comment]
For anyone who might be interested in pure C++ implementation of this tutorial:
https://github.com/chuzcjoe/CORE/blob/master/examples/metal/MetalTriangleDemo/main.cpp
